core/utils/historied-data/src/linear.rs

```diff
@@ -0,0 +1,515 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Transactional overlay implementation.
+//!
+//! This follows a linear succession of states.
+//! This contains multiple unbounded transaction layer
+//! and an additional top level 'prospective' layer.
+//! It only allows linear history (no branch so
+//! inner storage is only an array of element).
+
+use crate::State as TransactionState;
+use rstd::vec::Vec;
+use rstd::vec;
+
+/// An entry at a given history height.
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+pub struct HistoriedValue<V> {
+	/// The stored value.
+	pub value: V,
+	/// The moment in history when the value got set.
+	pub index: usize,
+}
+
+impl<V> From<(V, usize)> for HistoriedValue<V> {
+	fn from(input: (V, usize)) -> HistoriedValue<V> {
+		HistoriedValue { value: input.0, index: input.1 }
+	}
+}
+
+impl<V> HistoriedValue<V> {
+	fn as_ref(&self) -> HistoriedValue<&V> {
+		HistoriedValue {
+			value: &self.value,
+			index: self.index,
+		}
+	}
+}
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(not(feature = "std"))]
+type MemoryOnly<V> = Vec<HistoriedValue<V>>;
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(feature = "std")]
+type MemoryOnly<V> = smallvec::SmallVec<[HistoriedValue<V>; ALLOCATED_HISTORY]>;
+
+/// Size of preallocated history per element.
+/// Currently at two for committed and prospective only.
+/// It means that using transaction in a module got a direct allocation cost.
+#[cfg(feature = "std")]
+const ALLOCATED_HISTORY: usize = 2;
+
+/// History of value that are related to a state history (eg field `history` of
+/// an `OverlayedChangeSet`).
+///
+/// Values are always paired with a state history index.
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+pub struct History<V>(MemoryOnly<V>);
+
+impl<V> Default for History<V> {
+	fn default() -> Self {
+		History(Default::default())
+	}
+}
+
+// Following implementation are here to isolate
+// buffer specific functions.
+impl<V> History<V> {
+
+	fn get_state(&self, index: usize) -> HistoriedValue<&V> {
+		self.0[index].as_ref()
+	}
+
+	#[cfg(any(test, feature = "test"))]
+	/// Create an history from an existing history.
+	pub fn from_iter(input: impl IntoIterator<Item = HistoriedValue<V>>) -> Self {
+		let mut history = History::default();
+		for v in input {
+			history.push_unchecked(v);
+		}
+		history
+	}
+
+	#[cfg(any(test, feature = "test"))]
+	pub fn len(&self) -> usize {
```

Comment:
Please remove the other implementation. Just make it public and good. It does not modify any state, so having it public will not harm anyone.